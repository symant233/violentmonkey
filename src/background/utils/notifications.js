import { i18n, defaultImage, sendTabCmd, trueJoin } from '@/common';
import { addPublicCommands, commands } from './init';
import { CHROME } from './ua';
import { vetUrl } from './url';

/** @type {{ [nid: string]: browser.runtime.MessageSender | function | number }} */
const openers = {};
const kZombie = 'zombie';
const kZombieTimeout = 'zombieTimeout';
const kZombieUrl = 'zombieUrl';

addPublicCommands({
  /** @return {Promise<string>} */
  async Notification({
    image,
    text,
    tag,
    title,
    silent,
    onclick,
    [kZombieUrl]: zombieUrl,
    [kZombieTimeout]: zombieTimeout,
  }, src) {
    if (tag) clearZombieTimer(openers[tag]);
    const notificationId = await browser.notifications.create(tag, {
      type: 'basic',
      title: [title, IS_FIREFOX && i18n('extName')]::trueJoin(' - '), // Chrome already shows the name
      message: text,
      iconUrl: image || defaultImage,
      ...!IS_FIREFOX && {
        requireInteraction: !!onclick,
      },
      ...CHROME >= 70 && {
        silent,
      }
    });
    if (isFunction(onclick)) {
      openers[notificationId] = onclick;
    } else if (src) {
      openers[notificationId] = src;
      if (+zombieTimeout > 0) src[kZombieTimeout] = +zombieTimeout;
      if (zombieUrl != null) src[kZombieUrl] = vetUrl(zombieUrl, src.url);
    }
    return notificationId;
  },
  RemoveNotification(nid) {
    clearZombieTimer(openers[nid]);
    removeNotification(nid);
  },
});

browser.notifications.onClicked.addListener((id) => {
  notifyOpener(id, true);
});

browser.notifications.onClosed.addListener((id) => {
  notifyOpener(id, false);
  delete openers[id];
});

function notifyOpener(id, isClick) {
  const op = openers[id];
  if (!op) return;
  if (isFunction(op)) {
    if (isClick) op();
  } else if (op > 0) {
    if (isClick) clearZombieTimer(op);
  } else if (op[kZombie]) {
    if (isClick) {
      commands.TabOpen({ url: op[kZombieUrl] }, op);
      removeNotification(id); // Chrome doesn't auto-remove it on click
    }
  } else {
    sendTabCmd(op.tab.id, isClick ? 'NotificationClick' : 'NotificationClose', id, {
      [kFrameId]: op[kFrameId],
    });
  }
}

export function clearNotifications(tabId, frameId, tabRemoved) {
  for (const nid in openers) {
    const op = openers[nid];
    if (isObject(op)
    && op.tab.id === tabId
    && (!frameId || op[kFrameId] === frameId)
    && !op[kZombie]) {
      if (op[kZombieTimeout]) {
        op[kZombie] = setTimeout(removeNotification, op[kZombieTimeout], nid);
        if (!op[kZombieUrl]) openers[nid] = op[kZombie];
        if (tabRemoved) op._removed = true;
      } else {
        removeNotification(nid);
      }
    }
  }
}

function clearZombieTimer(op) {
  if (op > 0) clearTimeout(op);
}

function removeNotification(nid) {
  delete openers[nid];
  return browser.notifications.clear(nid);
}
