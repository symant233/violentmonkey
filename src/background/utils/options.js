import { debounce, initHooks, normalizeKeys, sendCmd } from '@/common';
import { deepCopy, deepEqual, objectGet, objectSet } from '@/common/object';
import defaults, { kScriptTemplate } from '@/common/options-defaults';
import { addOwnCommands, init } from './init';
import storage from './storage';

let changes;

addOwnCommands({
  /** @return {Object} */
  GetAllOptions() {
    return Object.assign({}, defaults, options); // eslint-disable-line no-use-before-define
  },
  /**
   * @param {{ [key:string]: PlainJSONValue }} data
   * @return {void}
   * @throws {?} hooks can throw after the option was set */
  SetOptions(data) {
    for (const key in data) setOption(key, data[key], true);
    callHooks(); // exceptions will be sent to the caller
  },
});

const options = {};
export const kOptions = 'options';
export const kVersion = 'version';
const TPL_OLD_VAL = `\
// ==UserScript==
// @name New Script
// @namespace Violentmonkey Scripts
// @match {{url}}
// @grant none
// ==/UserScript==
`;
const DELAY = 100;
const hooks = initHooks();
const callHooksLater = debounce(callHooks, DELAY);
const writeOptionsLater = debounce(writeOptions, DELAY);
const optProxy = new Proxy(defaults, { get: (_, key) => getOption(key) });
export const hookOptions = hooks.hook;
hookOptions(data => sendCmd('UpdateOptions', data));

export function initOptions(data) {
  data = data[kOptions] || {};
  Object.assign(options, data);
  if (process.env.DEBUG) console.info('options:', options);
  if (!options[kVersion]) {
    setOption(kVersion, 1);
  }
  if (options[kScriptTemplate] === TPL_OLD_VAL) {
    options[kScriptTemplate] = defaults[kScriptTemplate]; // will be detected by omitDefaultValue below
  }
  if (Object.keys(options).map(omitDefaultValue).some(Boolean)) {
    delete options[`${kScriptTemplate}Edited`]; // TODO: remove this in 2023
    writeOptionsLater();
  }
}

/**
 * @param {!string} key - must be "a.b.c" to allow clients easily set inside existing object trees
 * @param {PlainJSONValue} [value]
 * @param {boolean} [silent] - in case you callHooks() directly yourself afterwards
 */
function addChange(key, value, silent) {
  if (!changes) changes = {};
  else delete changes[key]; // Deleting first to place the new value at the end
  changes[key] = value;
  if (!silent) callHooksLater();
}

/** @throws in option handlers */
function callHooks() {
  if (!changes) return; // may happen in callHooksLater if callHooks was called earlier
  const tmp = changes;
  changes = null;
  hooks.fire(tmp);
}

/** Hooks and calls the callback with a copy of all options when init is resolved */
export function hookOptionsInit(cb) {
  if (init) init.then(() => cb(optProxy, true));
  else cb(optProxy, true);
  return hookOptions(cb);
}

export function getOption(key) {
  let res = options[key];
  if (res != null) return res;
  const keys = normalizeKeys(key);
  const mainKey = keys[0];
  const value = options[mainKey] ?? deepCopy(defaults[mainKey]);
  return keys.length > 1 ? objectGet(value, keys.slice(1)) : value;
}

export function setOption(key, value, silent) {
  if (init) return init.then(setOption.bind(null, ...arguments));
  const keys = normalizeKeys(key);
  const mainKey = keys[0];
  key = keys.join('.'); // must be a string for addChange()
  if (!hasOwnProperty(defaults, mainKey)) {
    if (process.env.DEBUG) console.info('Unknown option:', key, value, options);
    return;
  }
  const subKey = keys.length > 1 && keys.slice(1);
  const mainVal = getOption([mainKey]);
  if (deepEqual(value, subKey ? objectGet(mainVal, subKey) : mainVal)) {
    if (process.env.DEBUG) console.info('Option unchanged:', key, value, options);
    return;
  }
  options[mainKey] = subKey ? objectSet(mainVal, subKey, value) : value;
  omitDefaultValue(mainKey);
  writeOptionsLater();
  addChange(key, value, silent);
  if (process.env.DEBUG) console.info('Options updated:', key, value, options);
}

function writeOptions() {
  return storage.base.setOne(kOptions, options);
}

function omitDefaultValue(key) {
  return deepEqual(options[key], defaults[key])
    && delete options[key];
}
