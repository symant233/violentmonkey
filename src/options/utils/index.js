import { reactive } from 'vue';
import { sendCmdDirectly } from '@/common';
import { route } from '@/common/router';

export * from './search';

export const store = reactive({
  route,
  batch: null,
  /** Speedup and deflicker initial page load by not rendering an invisible script list */
  canRenderScripts: [SCRIPTS, TAB_RECYCLE, ''].includes(route.hash),
  scripts: [],
  removedScripts: [],
  loading: false,
  /** Whether removed scripts need to be filtered from `store.scripts`. */
  needRefresh: false,
  sync: [],
  title: null,
});

export const kInclude = 'include';
export const kMatch = 'match';
export const kExclude = 'exclude';
export const kExcludeMatch = 'excludeMatch';
export const kDescription = 'description';
export const kDownloadURL = 'downloadURL';
export const kHomepageURL = 'homepageURL';
export const kIcon = 'icon';
export const kName = 'name';
export const kOrigExclude = 'origExclude';
export const kOrigExcludeMatch = 'origExcludeMatch';
export const kOrigInclude = 'origInclude';
export const kOrigMatch = 'origMatch';
export const kStorageSize = 'storageSize';
export const kUpdateURL = 'updateURL';

export let K_SAVE; // deduced from the current CodeMirror keymap

export function inferSaveHotKey(hotkeys) {
  K_SAVE = hotkeys.find(([, cmd]) => cmd === 'save')?.[0];
  if (!K_SAVE) {
    K_SAVE = 'Ctrl-S';
    hotkeys.unshift([K_SAVE, 'save']);
  }
}

export function markRemove(script, removed) {
  return sendCmdDirectly('MarkRemoved', {
    id: script.props.id,
    removed,
  });
}

export async function runInBatch(fn, ...args) {
  try {
    await (store.batch = fn(...args) || true);
  } finally {
    store.batch = false;
  }
}

export function toggleBoolean(event) {
  const el = /** @type {HTMLTextAreaElement} */ event.target;
  const { selectionStart: start, selectionEnd: end, value } = el;
  // Ignoring double-clicks outside of <textarea>
  const toggled = end && { false: 'true', true: 'false' }[value.slice(start, end)];
  // FF can't run execCommand on textarea, https://bugzil.la/1220696#c24
  if (toggled && !document.execCommand('insertText', false, toggled)) {
    el.value = value.slice(0, start) + toggled + value.slice(end);
    el.setSelectionRange(start + toggled.length, start + toggled.length);
    el.dispatchEvent(new Event('input'));
    el.onblur = () => el.dispatchEvent(new Event('change'));
  }
}
