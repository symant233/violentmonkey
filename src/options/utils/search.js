import { escapeStringForRegExp, normalizeTag } from '@/common';

const reToken = re`/\s*
  (!)?
  (
    \# |
    (name|code|desc)(\+re)?: |
    (re:) |
  )
  (
    '((?:[^']+|'')*) ('|$) |
    "((?:[^"]+|"")*) ("|$) |
    \/(\S+?)\/([a-z]*) |
    \S+
  )
  (?:\s+|$)
/yx`;
const reTwoSingleQuotes = /''/g;
const reTwoDoubleQuotes = /""/g;

export function createSearchRules(search) {
  /** @type {VMSearchRule[]} */
  const rules = [];
  const tokens = [];
  const includeTags = [];
  const excludeTags = [];
  reToken.lastIndex = 0;
  for (let m; (m = reToken.exec(search)); ) {
    let [,
      negative,
      prefix, scope = '', re1, re2,
      raw,
      q1, q1end,
      quoted = q1, quoteEnd = q1end,
      reStr, flags = '',
    ] = m;
    let str;
    if (quoted) {
      if (!quoteEnd) throw new Error('Unmatched quotes');
      str = quoted.replace(q1 ? reTwoSingleQuotes : reTwoDoubleQuotes, quoted[0]);
    } else {
      str = raw;
    }
    negative = !!negative;
    tokens.push({
      negative,
      prefix,
      raw,
      parsed: str,
    });
    if (prefix === '#') {
      str = normalizeTag(str).replace(/\./g, '\\.');
      if (str) (negative ? excludeTags : includeTags).push(str);
    } else {
      if (re1 || re2) {
        flags = 'i';
      } else if (reStr) {
        str = reStr;
      } else {
        if (!quoted) flags = 'i';
        str = escapeStringForRegExp(str);
      }
      /** @namespace VMSearchRule */
      rules.push({
        negative,
        scope,
        re: new RegExp(str, flags.includes('u') ? flags : flags + 'u'),
      });
    }
  }
  [includeTags, excludeTags].forEach((tags, negative) => {
    if (tags.length) {
      rules.unshift({
        scope: 'tags',
        re: new RegExp(`(?:^|\\s)(${tags.join('|').toLowerCase()})(\\s|$)`, 'u'),
        negative: !!negative,
      });
    }
  });
  return {
    tokens,
    rules,
  };
}

/**
 * @this {Object} $cache, see initScript()
 * @param {VMSearchRule} rule
 * @return {number}
 */
export function testSearchRule({ re, negative, scope }) {
  return negative ^ (
    re.test(this[scope || 'desc'])
    || !scope && re.test(this.code)
  );
}

export function performSearch(scripts, rules) {
  let res = 0;
  for (const { $cache } of scripts) {
    res += ($cache.show = rules.every(testSearchRule, $cache));
  }
  return res;
}
