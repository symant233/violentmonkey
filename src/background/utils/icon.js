import { i18n, ignoreChromeErrors, makeDataUri, noop } from '@/common';
import { BLACKLIST } from '@/common/consts';
import { nest, objectPick } from '@/common/object';
import { addOwnCommands, commands, init } from './init';
import { getOption, hookOptions, setOption } from './options';
import { popupTabs } from './popup-tracker';
import storage, { S_CACHE } from './storage';
import { forEachTab, getTabUrl, injectableRe, openDashboard, tabsOnRemoved, tabsOnUpdated } from './tabs';
import { testBlacklist } from './tester';
import { FIREFOX } from './ua';

/** We don't set 19px because FF and Vivaldi scale it down to 16px instead of our own crisp 16px */
const SIZES = [16, 32];
/** Caching own icon to improve dashboard loading speed, as well as browserAction API
 * (e.g. Chrome wastes 40ms in our extension's process to read 4 icons for every tab). */
const iconCache = {};
const iconDataCache = {};
/** @return {string | Promise<string>} */
export const getImageData = url => iconCache[url] || (iconCache[url] = loadIcon(url));
// Firefox Android does not support such APIs, use noop
const browserAction = (() => {
  // Using `chrome` namespace in order to skip our browser.js polyfill in Chrome
  const api = chrome.browserAction;
  // Some methods like setBadgeText added callbacks only in Chrome 67+.
  const makeMethod = fn => (...args) => {
    try {
      // Suppress the "no tab id" error when setting an icon/badge as it cannot be reliably prevented
      api::fn(...args, ignoreChromeErrors);
    } catch (e) {
      api::fn(...args);
    }
  };
  return objectPick(api, [
    'setIcon',
    'setBadgeText',
    'setBadgeBackgroundColor',
    'setTitle',
  ], fn => (fn ? makeMethod(fn) : noop));
})();
// Promisifying explicitly because this API returns an id in Firefox and not a Promise
const contextMenus = chrome.contextMenus;

/** @type {{ [tabId: string]: VMBadgeData }}*/
export const badges = {};
const KEY_SHOW_BADGE = 'showBadge';
const KEY_BADGE_COLOR = 'badgeColor';
const KEY_BADGE_COLOR_BLOCKED = 'badgeColorBlocked';
const titleBlacklisted = i18n('failureReasonBlacklisted');
const titleDefault = extensionManifest[BROWSER_ACTION].default_title;
const iconDefault = extensionManifest[BROWSER_ACTION].default_icon[16].match(/\d+(\w*)\./)[1];
const titleDisabled = i18n('menuScriptDisabled');
const titleNoninjectable = i18n('failureReasonNoninjectable');
const titleSkipped = i18n('skipScriptsMsg');
let isApplied;
/** @type {VMBadgeMode} */
let showBadge;
let badgeColor;
let badgeColorBlocked;

addOwnCommands({
  GetImageData: getImageData,
});

hookOptions((changes) => {
  let v;
  const jobs = [];
  if ((v = changes[IS_APPLIED]) != null) {
    isApplied = v;
    setIcon(); // change the default icon
    jobs.push(setIcon); // change the current tabs' icons
  }
  if ((v = changes[KEY_SHOW_BADGE]) != null) {
    showBadge = v;
    jobs.push(updateBadge);
    contextMenus?.update(KEY_SHOW_BADGE + ':' + showBadge, {checked: true});
  }
  if ((v = changes[KEY_BADGE_COLOR]) && (badgeColor = v)
  || (v = changes[KEY_BADGE_COLOR_BLOCKED]) && (badgeColorBlocked = v)) {
    jobs.push(updateBadgeColor);
  }
  if (BLACKLIST in changes) {
    jobs.push(updateState);
  }
  if (jobs.length) {
    forEachTab(tab => jobs.forEach(fn => fn(tab)));
  }
});

init.then(async () => {
  isApplied = getOption(IS_APPLIED);
  showBadge = getOption(KEY_SHOW_BADGE);
  badgeColor = getOption(KEY_BADGE_COLOR);
  badgeColorBlocked = getOption(KEY_BADGE_COLOR_BLOCKED);
  forEachTab(updateState);
  if (!isApplied) setIcon(); // sets the dimmed icon as default
  if (contextMenus) {
    const addToIcon = (id, title, opts) => (
      new Promise(resolve => (
        contextMenus.create({
          contexts: [BROWSER_ACTION],
          id,
          title,
          ...opts,
        }, resolve)
      ))
    ).then(ignoreChromeErrors);
    const badgeChild = { parentId: KEY_SHOW_BADGE, type: 'radio' };
    await addToIcon(SKIP_SCRIPTS, i18n('skipScripts'));
    for (const args of [
      [KEY_SHOW_BADGE, i18n('labelBadge')],
      [`${KEY_SHOW_BADGE}:`, i18n('labelBadgeNone'), badgeChild],
      [`${KEY_SHOW_BADGE}:unique`, i18n('labelBadgeUnique'), badgeChild],
      [`${KEY_SHOW_BADGE}:total`, i18n('labelBadgeTotal'), badgeChild],
    ]) {
      await addToIcon(...args);
    }
    contextMenus.update(KEY_SHOW_BADGE + ':' + showBadge, { checked: true });
    // Chrome already adds a built-in "Options" item
    if (IS_FIREFOX) await addToIcon(TAB_SETTINGS, i18n('labelSettings'));
  }
});

contextMenus?.onClicked.addListener(({ menuItemId: id }, tab) => {
  handleHotkeyOrMenu(id, tab);
});
tabsOnRemoved.addListener(id => delete badges[id]);
tabsOnUpdated.addListener((tabId, { url }, tab) => {
  if (url) {
    const [title] = getFailureReason(url);
    if (title) updateState(tab, resetBadgeData(tabId, null), title);
  }
}, FIREFOX && { properties: ['status'] });

function resetBadgeData(tabId, isInjected) {
  // 'total' and 'unique' must match showBadge in options-defaults.js
  /** @type {VMBadgeData} */
  const data = nest(badges, tabId);
  data.icon = iconDefault;
  data.total = 0;
  data.unique = 0;
  data[IDS] = new Set();
  data[kFrameId] = undefined;
  data[INJECT] = isInjected;
  // Notify popup about non-injectable tab
  if (!isInjected) popupTabs[tabId]?.postMessage(null);
  return data;
}

/**
 * @param {number[] | string} ids
 * @param {boolean} reset
 * @param {VMMessageSender} src
 */
export function setBadge(ids, reset, { tab, [kFrameId]: frameId, [kTop]: isTop }) {
  const tabId = tab.id;
  const injectable = ids === SKIP_SCRIPTS || ids === 'off' ? ids : !!ids;
  /** @type {VMBadgeData} */
  const data = !(reset && isTop) && badges[tabId] || resetBadgeData(tabId, injectable);
  if (Array.isArray(ids)) {
    const {
      [IDS]: idMap,
      [kFrameId]: totalMap = data[kFrameId] = {},
    } = data;
    // uniques
    ids.forEach(idMap.add, idMap);
    data.unique = idMap.size;
    // totals
    data.total = 0;
    totalMap[frameId] = ids.length;
    for (const id in totalMap) data.total += totalMap[id];
  }
  if (isTop) {
    data[INJECT] = injectable;
  }
  updateBadgeColor(tab, data);
  updateState(tab, data);
}

function updateBadge({ id: tabId }, data = badges[tabId]) {
  if (data) {
    browserAction.setBadgeText({
      text: `${data[showBadge] || ''}`,
      tabId,
    });
  }
}

function updateBadgeColor({ id: tabId }, data = badges[tabId]) {
  if (data) {
    browserAction.setBadgeBackgroundColor({
      color: data[INJECT] ? badgeColor : badgeColorBlocked,
      tabId,
    });
  }
}

function updateState(tab, data, title) {
  const tabId = tab.id;
  if (!data) data = badges[tabId] || resetBadgeData(tabId);
  if (!title) [title] = getFailureReason(getTabUrl(tab), data);
  browserAction.setTitle({ tabId, title });
  setIcon(tab, data);
  updateBadge(tab, data);
}

async function setIcon({ id: tabId } = {}, data = badges[tabId] || {}) {
  const mod = !isApplied ? 'w'
    : data[INJECT] !== true ? 'b'
      : '';
  if (data.icon === mod) return;
  data.icon = mod;
  const pathData = {};
  const iconData = {};
  for (const n of SIZES) {
    const url = `${ICON_PREFIX}${n}${mod}.png`;
    pathData[n] = url;
    iconData[n] = iconDataCache[url]
      || await (iconCache[url] || (iconCache[url] = loadIcon(url))) && iconDataCache[url];
  }
  // imageData doesn't work in Firefox Android, so we also set path here
  browserAction.setIcon({
    tabId,
    path: pathData,
    imageData: iconData,
  });
}

/** Omitting `data` = check whether injection is allowed for `url` */
export function getFailureReason(url, data, def = titleDefault) {
  return !injectableRe.test(url) ? [titleNoninjectable, INJECT_INTO]
    : ((url = testBlacklist(url))) ? [titleBlacklisted, 'blacklisted', url]
      : !isApplied || data?.[INJECT] === 'off' ? [titleDisabled, IS_APPLIED]
        : !data ? []
          : data[INJECT] === SKIP_SCRIPTS
            ? [titleSkipped, SKIP_SCRIPTS]
            : [def];
}

export function handleHotkeyOrMenu(id, tab) {
  if (id === SKIP_SCRIPTS) {
    commands[SKIP_SCRIPTS](tab);
  } else if (id === TAB_SETTINGS) {
    openDashboard(id);
  } else if (id === 'dashboard') {
    openDashboard('');
  } else if (id === 'newScript') {
    commands.OpenEditor();
  } else if (id === 'updateScripts') {
    commands.CheckUpdate();
  } else if (id === 'updateScriptsInTab') {
    id = badges[tab.id]?.[IDS];
    if (id) commands.CheckUpdate([...id]);
  } else if (id.startsWith(KEY_SHOW_BADGE)) {
    setOption(KEY_SHOW_BADGE, id.slice(KEY_SHOW_BADGE.length + 1));
  }
}

async function loadIcon(url) {
  const img = new Image();
  const isOwn = url.startsWith(ICON_PREFIX);
  img.src = isOwn ? url.slice(extensionOrigin.length) // must be a relative path in Firefox Android
    : url.startsWith('data:') ? url
      : makeDataUri(url[0] === 'i' ? url : await loadStorageCache(url));
  await new Promise((resolve) => {
    img.onload = resolve;
    img.onerror = resolve;
  });
  let res;
  let maxSize = !isOwn && (2 * 38); // dashboard icon size for 2xDPI
  let { width, height } = img;
  if (!width || !height) { // FF reports 0 for SVG
    iconCache[url] = url;
    return url;
  }
  if (maxSize && (width > maxSize || height > maxSize)) {
    maxSize /= width > height ? width : height;
    width = Math.round(width * maxSize);
    height = Math.round(height * maxSize);
  }
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');
  canvas.width = width;
  canvas.height = height;
  ctx.drawImage(img, 0, 0, width, height);
  try {
    res = canvas.toDataURL();
    if (isOwn) iconDataCache[url] = ctx.getImageData(0, 0, width, height);
  } catch (err) {
    res = url;
  }
  iconCache[url] = res;
  return res;
}

async function loadStorageCache(url) {
  return await storage[S_CACHE].getOne(url)
    ?? await storage[S_CACHE].fetch(url, 'res').catch(console.warn);
}
