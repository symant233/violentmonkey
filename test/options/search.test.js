import { createSearchRules } from '@/options/utils/search';

test('createSearchRules', () => {
  expect(createSearchRules('')).toMatchSnapshot();
  expect(createSearchRules('#a #b !#c hello "CaseSensitive" CaseInsensitive')).toMatchSnapshot();
  expect(createSearchRules('#a-b #b name:hello world')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b name:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b name+re:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('#a.b #b !name+re:"hello world"')).toMatchSnapshot();
  expect(createSearchRules('"#a.b" !"#b"')).toMatchSnapshot();
  expect(createSearchRules(String.raw`/\d+\D+/ code:/\d+\D+/u /not regexp/`)).toMatchSnapshot();
  expect(createSearchRules('foobar re:foobar name+re:foobar code+re:foobar')).toMatchSnapshot();
});
