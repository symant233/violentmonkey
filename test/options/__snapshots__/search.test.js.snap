// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSearchRules 1`] = `
{
  "rules": [],
  "tokens": [],
}
`;

exports[`createSearchRules 2`] = `
{
  "rules": [
    {
      "negative": true,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(c\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": false,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(a\\|b\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": false,
      "re": /hello/iu,
      "scope": "",
    },
    {
      "negative": false,
      "re": /CaseSensitive/u,
      "scope": "",
    },
    {
      "negative": false,
      "re": /CaseInsensitive/iu,
      "scope": "",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "a",
      "prefix": "#",
      "raw": "a",
    },
    {
      "negative": false,
      "parsed": "b",
      "prefix": "#",
      "raw": "b",
    },
    {
      "negative": true,
      "parsed": "c",
      "prefix": "#",
      "raw": "c",
    },
    {
      "negative": false,
      "parsed": "hello",
      "prefix": "",
      "raw": "hello",
    },
    {
      "negative": false,
      "parsed": "CaseSensitive",
      "prefix": "",
      "raw": ""CaseSensitive"",
    },
    {
      "negative": false,
      "parsed": "CaseInsensitive",
      "prefix": "",
      "raw": "CaseInsensitive",
    },
  ],
}
`;

exports[`createSearchRules 3`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(a-b\\|b\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": false,
      "re": /hello/iu,
      "scope": "name",
    },
    {
      "negative": false,
      "re": /world/iu,
      "scope": "",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "a-b",
      "prefix": "#",
      "raw": "a-b",
    },
    {
      "negative": false,
      "parsed": "b",
      "prefix": "#",
      "raw": "b",
    },
    {
      "negative": false,
      "parsed": "hello",
      "prefix": "name:",
      "raw": "hello",
    },
    {
      "negative": false,
      "parsed": "world",
      "prefix": "",
      "raw": "world",
    },
  ],
}
`;

exports[`createSearchRules 4`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": false,
      "re": /hello world/u,
      "scope": "name",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "a.b",
      "prefix": "#",
      "raw": "a.b",
    },
    {
      "negative": false,
      "parsed": "b",
      "prefix": "#",
      "raw": "b",
    },
    {
      "negative": false,
      "parsed": "hello world",
      "prefix": "name:",
      "raw": ""hello world"",
    },
  ],
}
`;

exports[`createSearchRules 5`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": false,
      "re": /hello world/iu,
      "scope": "name",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "a.b",
      "prefix": "#",
      "raw": "a.b",
    },
    {
      "negative": false,
      "parsed": "b",
      "prefix": "#",
      "raw": "b",
    },
    {
      "negative": false,
      "parsed": "hello world",
      "prefix": "name+re:",
      "raw": ""hello world"",
    },
  ],
}
`;

exports[`createSearchRules 6`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /\\(\\?:\\^\\|\\\\s\\)\\(a\\\\\\.b\\|b\\)\\(\\\\s\\|\\$\\)/u,
      "scope": "tags",
    },
    {
      "negative": true,
      "re": /hello world/iu,
      "scope": "name",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "a.b",
      "prefix": "#",
      "raw": "a.b",
    },
    {
      "negative": false,
      "parsed": "b",
      "prefix": "#",
      "raw": "b",
    },
    {
      "negative": true,
      "parsed": "hello world",
      "prefix": "name+re:",
      "raw": ""hello world"",
    },
  ],
}
`;

exports[`createSearchRules 7`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /#a\\\\\\.b/u,
      "scope": "",
    },
    {
      "negative": true,
      "re": /#b/u,
      "scope": "",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "#a.b",
      "prefix": "",
      "raw": ""#a.b"",
    },
    {
      "negative": true,
      "parsed": "#b",
      "prefix": "",
      "raw": ""#b"",
    },
  ],
}
`;

exports[`createSearchRules 8`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /\\\\d\\+\\\\D\\+/u,
      "scope": "",
    },
    {
      "negative": false,
      "re": /\\\\d\\+\\\\D\\+/u,
      "scope": "code",
    },
    {
      "negative": false,
      "re": /\\\\/not/iu,
      "scope": "",
    },
    {
      "negative": false,
      "re": /regexp\\\\//iu,
      "scope": "",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "/\\d+\\D+/",
      "prefix": "",
      "raw": "/\\d+\\D+/",
    },
    {
      "negative": false,
      "parsed": "/\\d+\\D+/u",
      "prefix": "code:",
      "raw": "/\\d+\\D+/u",
    },
    {
      "negative": false,
      "parsed": "/not",
      "prefix": "",
      "raw": "/not",
    },
    {
      "negative": false,
      "parsed": "regexp/",
      "prefix": "",
      "raw": "regexp/",
    },
  ],
}
`;

exports[`createSearchRules 9`] = `
{
  "rules": [
    {
      "negative": false,
      "re": /foobar/iu,
      "scope": "",
    },
    {
      "negative": false,
      "re": /foobar/iu,
      "scope": "",
    },
    {
      "negative": false,
      "re": /foobar/iu,
      "scope": "name",
    },
    {
      "negative": false,
      "re": /foobar/iu,
      "scope": "code",
    },
  ],
  "tokens": [
    {
      "negative": false,
      "parsed": "foobar",
      "prefix": "",
      "raw": "foobar",
    },
    {
      "negative": false,
      "parsed": "foobar",
      "prefix": "re:",
      "raw": "foobar",
    },
    {
      "negative": false,
      "parsed": "foobar",
      "prefix": "name+re:",
      "raw": "foobar",
    },
    {
      "negative": false,
      "parsed": "foobar",
      "prefix": "code+re:",
      "raw": "foobar",
    },
  ],
}
`;
